In [ ]:
import numpy as np
import pandas as pd

df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/mpg (1).csv")


In [ ]:
df

mpg  cylinders  displacement horsepower  weight  acceleration  \
0    18.0          8         307.0        130    3504          12.0   
1    15.0          8         350.0        165    3693          11.5   
2    18.0          8         318.0        150    3436          11.0   
3    16.0          8         304.0        150    3433          12.0   
4    17.0          8         302.0        140    3449          10.5   
..    ...        ...           ...        ...     ...           ...   
393  27.0          4         140.0         86    2790          15.6   
394  44.0          4          97.0         52    2130          24.6   
395  32.0          4         135.0         84    2295          11.6   
396  28.0          4         120.0         79    2625          18.6   
397  31.0          4         119.0         82    2720          19.4   

     model_year  origin                       name  
0            70       1  chevrolet chevelle malibu  
1            70       1          buick skylark 320  
2            70       1         plymouth satellite  
3            70       1              amc rebel sst  
4            70       1                ford torino  
..          ...     ...                        ...  
393          82       1            ford mustang gl  
394          82       2                  vw pickup  
395          82       1              dodge rampage  
396          82       1                ford ranger  
397          82       1                 chevy s-10  

[398 rows x 9 columns]

In [ ]:
# Groupby Basics
df['model_year'].unique()

array([70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82])

In [ ]:
df.groupby('model_year') # cannot be group by just like this, has to use mean, count, etc

In [ ]:
g_mean = df.groupby('model_year').mean(numeric_only= True)

In [ ]:
g_mean

mpg  cylinders  displacement       weight  acceleration  \
model_year                                                                  
70          17.689655   6.758621    281.413793  3372.793103     12.948276   
71          21.250000   5.571429    209.750000  2995.428571     15.142857   
72          18.714286   5.821429    218.375000  3237.714286     15.125000   
73          17.100000   6.375000    256.875000  3419.025000     14.312500   
74          22.703704   5.259259    171.740741  2877.925926     16.203704   
75          20.266667   5.600000    205.533333  3176.800000     16.050000   
76          21.573529   5.647059    197.794118  3078.735294     15.941176   
77          23.375000   5.464286    191.392857  2997.357143     15.435714   
78          24.061111   5.361111    177.805556  2861.805556     15.805556   
79          25.093103   5.827586    206.689655  3055.344828     15.813793   
80          33.696552   4.137931    115.827586  2436.655172     16.934483   
81          30.334483   4.620690    135.310345  2522.931034     16.306897   
82          31.709677   4.193548    128.870968  2453.548387     16.638710   

              origin  
model_year            
70          1.310345  
71          1.428571  
72          1.535714  
73          1.375000  
74          1.666667  
75          1.466667  
76          1.470588  
77          1.571429  
78          1.611111  
79          1.275862  
80          2.206897  
81          1.965517  
82          1.645161

In [ ]:
g_displacement = df.groupby('displacement').mean(numeric_only= True)

In [ ]:
g_displacement

mpg  cylinders       weight  acceleration  model_year  \
displacement                                                                
68.0          29.000000        4.0  1867.000000     19.500000   73.000000   
70.0          20.233333        3.0  2291.333333     13.166667   75.000000   
71.0          31.500000        4.0  1804.500000     20.000000   72.500000   
72.0          35.000000        4.0  1613.000000     18.000000   71.000000   
76.0          31.000000        4.0  1649.000000     16.500000   74.000000   
...                 ...        ...          ...           ...         ...   
400.0         13.961538        8.0  4521.230769     11.715385   72.923077   
429.0         12.666667        8.0  4642.000000     10.833333   71.666667   
440.0         13.500000        8.0  4523.500000      9.750000   71.500000   
454.0         14.000000        8.0  4354.000000      9.000000   70.000000   
455.0         13.333333        8.0  4154.000000     10.333333   71.000000   

              origin  
displacement          
68.0             2.0  
70.0             3.0  
71.0             3.0  
72.0             3.0  
76.0             3.0  
...              ...  
400.0            1.0  
429.0            1.0  
440.0            1.0  
454.0            1.0  
455.0            1.0  

[82 rows x 6 columns]

In [ ]:
# 1.1 GroupBy via Count

In [ ]:
g_count = df.groupby('cylinders').count()


In [ ]:
g_count

mpg  displacement  horsepower  weight  acceleration  model_year  \
cylinders                                                                    
3            4             4           4       4             4           4   
4          204           204         204     204           204         204   
5            3             3           3       3             3           3   
6           84            84          84      84            84          84   
8          103           103         103     103           103         103   

           origin  name  
cylinders                
3               4     4  
4             204   204  
5               3     3  
6              84    84  
8             103   103

In [ ]:
# 1.2 GroupBy via Multiple Column

In [ ]:
df

mpg  cylinders  displacement horsepower  weight  acceleration  \
0    18.0          8         307.0        130    3504          12.0   
1    15.0          8         350.0        165    3693          11.5   
2    18.0          8         318.0        150    3436          11.0   
3    16.0          8         304.0        150    3433          12.0   
4    17.0          8         302.0        140    3449          10.5   
..    ...        ...           ...        ...     ...           ...   
393  27.0          4         140.0         86    2790          15.6   
394  44.0          4          97.0         52    2130          24.6   
395  32.0          4         135.0         84    2295          11.6   
396  28.0          4         120.0         79    2625          18.6   
397  31.0          4         119.0         82    2720          19.4   

     model_year  origin                       name  
0            70       1  chevrolet chevelle malibu  
1            70       1          buick skylark 320  
2            70       1         plymouth satellite  
3            70       1              amc rebel sst  
4            70       1                ford torino  
..          ...     ...                        ...  
393          82       1            ford mustang gl  
394          82       2                  vw pickup  
395          82       1              dodge rampage  
396          82       1                ford ranger  
397          82       1                 chevy s-10  

[398 rows x 9 columns]

In [ ]:
df.groupby(['model_year', 'cylinders']).mean(numeric_only= True)

mpg  displacement       weight  acceleration  \
model_year cylinders                                                       
70         4          25.285714    107.000000  2292.571429     16.000000   
           6          20.500000    199.000000  2710.500000     15.500000   
           8          14.111111    367.555556  3940.055556     11.194444   
71         4          27.461538    101.846154  2056.384615     16.961538   
           6          18.000000    243.375000  3171.875000     14.750000   
           8          13.428571    371.714286  4537.714286     12.214286   
72         3          19.000000     70.000000  2330.000000     13.500000   
           4          23.428571    111.535714  2382.642857     17.214286   
           8          13.615385    344.846154  4228.384615     13.000000   
73         3          18.000000     70.000000  2124.000000     13.500000   
           4          22.727273    109.272727  2338.090909     17.136364   
           6          19.000000    212.250000  2917.125000     15.687500   
           8          13.200000    365.250000  4279.050000     12.250000   
74         4          27.800000     96.533333  2151.466667     16.400000   
           6          17.857143    230.428571  3320.000000     16.857143   
           8          14.200000    315.200000  4438.400000     14.700000   
75         4          25.250000    114.833333  2489.250000     15.833333   
           6          17.583333    233.750000  3398.333333     17.708333   
           8          15.666667    330.500000  4108.833333     13.166667   
76         4          26.766667    106.333333  2306.600000     16.866667   
           6          20.000000    221.400000  3349.600000     17.000000   
           8          14.666667    324.000000  4064.666667     13.222222   
77         3          21.500000     80.000000  2720.000000     13.500000   
           4          29.107143    106.500000  2205.071429     16.064286   
           6          19.500000    220.400000  3383.000000     16.900000   
           8          16.000000    335.750000  4177.500000     13.662500   
78         4          29.576471    112.117647  2296.764706     16.282353   
           5          20.300000    131.000000  2830.000000     15.900000   
           6          19.066667    213.250000  3314.166667     16.391667   
           8          19.050000    300.833333  3563.333333     13.266667   
79         4          31.525000    113.583333  2357.583333     15.991667   
           5          25.400000    183.000000  3530.000000     20.100000   
           6          22.950000    205.666667  3025.833333     15.433333   
           8          18.630000    321.400000  3862.900000     15.400000   
80         3          23.700000     70.000000  2420.000000     12.500000   
           4          34.612000    111.000000  2360.080000     17.144000   
           5          36.400000    121.000000  2950.000000     19.900000   
           6          25.900000    196.500000  3145.500000     15.050000   
81         4          32.814286    108.857143  2275.476190     16.466667   
           6          23.428571    184.000000  3093.571429     15.442857   
           8          26.600000    350.000000  3725.000000     19.000000   
82         4          32.071429    118.571429  2402.321429     16.703571   
           6          28.333333    225.000000  2931.666667     16.033333   

                        origin  
model_year cylinders            
70         4          2.285714  
           6          1.000000  
           8          1.000000  
71         4          1.923077  
           6          1.000000  
           8          1.000000  
72         3          3.000000  
           4          1.928571  
           8          1.000000  
73         3          3.000000  
           4          2.000000  
           6          1.250000  
           8          1.000000  
74         4          2.200000  
           6          1.000000  
           8          1.000000  
75         4 

In [ ]:
df.groupby(['cylinders', 'model_year']).mean(numeric_only= True)

mpg  displacement       weight  acceleration  \
cylinders model_year                                                       
3         72          19.000000     70.000000  2330.000000     13.500000   
          73          18.000000     70.000000  2124.000000     13.500000   
          77          21.500000     80.000000  2720.000000     13.500000   
          80          23.700000     70.000000  2420.000000     12.500000   
4         70          25.285714    107.000000  2292.571429     16.000000   
          71          27.461538    101.846154  2056.384615     16.961538   
          72          23.428571    111.535714  2382.642857     17.214286   
          73          22.727273    109.272727  2338.090909     17.136364   
          74          27.800000     96.533333  2151.466667     16.400000   
          75          25.250000    114.833333  2489.250000     15.833333   
          76          26.766667    106.333333  2306.600000     16.866667   
          77          29.107143    106.500000  2205.071429     16.064286   
          78          29.576471    112.117647  2296.764706     16.282353   
          79          31.525000    113.583333  2357.583333     15.991667   
          80          34.612000    111.000000  2360.080000     17.144000   
          81          32.814286    108.857143  2275.476190     16.466667   
          82          32.071429    118.571429  2402.321429     16.703571   
5         78          20.300000    131.000000  2830.000000     15.900000   
          79          25.400000    183.000000  3530.000000     20.100000   
          80          36.400000    121.000000  2950.000000     19.900000   
6         70          20.500000    199.000000  2710.500000     15.500000   
          71          18.000000    243.375000  3171.875000     14.750000   
          73          19.000000    212.250000  2917.125000     15.687500   
          74          17.857143    230.428571  3320.000000     16.857143   
          75          17.583333    233.750000  3398.333333     17.708333   
          76          20.000000    221.400000  3349.600000     17.000000   
          77          19.500000    220.400000  3383.000000     16.900000   
          78          19.066667    213.250000  3314.166667     16.391667   
          79          22.950000    205.666667  3025.833333     15.433333   
          80          25.900000    196.500000  3145.500000     15.050000   
          81          23.428571    184.000000  3093.571429     15.442857   
          82          28.333333    225.000000  2931.666667     16.033333   
8         70          14.111111    367.555556  3940.055556     11.194444   
          71          13.428571    371.714286  4537.714286     12.214286   
          72          13.615385    344.846154  4228.384615     13.000000   
          73          13.200000    365.250000  4279.050000     12.250000   
          74          14.200000    315.200000  4438.400000     14.700000   
          75          15.666667    330.500000  4108.833333     13.166667   
          76          14.666667    324.000000  4064.666667     13.222222   
          77          16.000000    335.750000  4177.500000     13.662500   
          78          19.050000    300.833333  3563.333333     13.266667   
          79          18.630000    321.400000  3862.900000     15.400000   
          81          26.600000    350.000000  3725.000000     19.000000   

                        origin  
cylinders model_year            
3         72          3.000000  
          73          3.000000  
          77          3.000000  
          80          3.000000  
4         70          2.285714  
          71          1.923077  
          72          1.928571  
          73          2.000000  
          74          2.200000  
          75          2.166667  
          76          1.866667  
          77          1.857143  
          78          2.117647  
          79          1.583333  
          80          2.200000  
          81          2.095238  
          82 

In [ ]:
#Example to sort descending
C_Groupby = df.groupby(['cylinders', 'model_year']).mean(numeric_only= True).head()
C_Groupby.sort_index(level = 'cylinders', ascending = False)

mpg  displacement       weight  acceleration  \
cylinders model_year                                                       
4         70          25.285714         107.0  2292.571429          16.0   
3         80          23.700000          70.0  2420.000000          12.5   
          77          21.500000          80.0  2720.000000          13.5   
          73          18.000000          70.0  2124.000000          13.5   
          72          19.000000          70.0  2330.000000          13.5   

                        origin  
cylinders model_year            
4         70          2.285714  
3         80          3.000000  
          77          3.000000  
          73          3.000000  
          72          3.000000

In [ ]:
C_Groupby.sort_index(level = 'model_year', ascending = False)

mpg  displacement       weight  acceleration  \
cylinders model_year                                                       
3         80          23.700000          70.0  2420.000000          12.5   
          77          21.500000          80.0  2720.000000          13.5   
          73          18.000000          70.0  2124.000000          13.5   
          72          19.000000          70.0  2330.000000          13.5   
4         70          25.285714         107.0  2292.571429          16.0   

                        origin  
cylinders model_year            
3         80          3.000000  
          77          3.000000  
          73          3.000000  
          72          3.000000  
4         70          2.285714

In [ ]:
C_Groupby.sort_index(level = ['model_year','cylinders'], ascending = [False, True])

mpg  displacement       weight  acceleration  \
cylinders model_year                                                       
3         80          23.700000          70.0  2420.000000          12.5   
          77          21.500000          80.0  2720.000000          13.5   
          73          18.000000          70.0  2124.000000          13.5   
          72          19.000000          70.0  2330.000000          13.5   
4         70          25.285714         107.0  2292.571429          16.0   

                        origin  
cylinders model_year            
3         80          3.000000  
          77          3.000000  
          73          3.000000  
          72          3.000000  
4         70          2.285714

In [ ]:
C_Groupby.sort_index(level = ['model_year','cylinders'], ascending = [True, False])

mpg  displacement       weight  acceleration  \
cylinders model_year                                                       
4         70          25.285714         107.0  2292.571429          16.0   
3         72          19.000000          70.0  2330.000000          13.5   
          73          18.000000          70.0  2124.000000          13.5   
          77          21.500000          80.0  2720.000000          13.5   
          80          23.700000          70.0  2420.000000          12.5   

                        origin  
cylinders model_year            
4         70          2.285714  
3         72          3.000000  
          73          3.000000  
          77          3.000000  
          80          3.000000

Exercise 1


1. How much beer is served on average in each continent
2. For each continent, print the overall statistics for wine consumption
3. Print the mean alcohol consumption per continent for every column
4. Print the median alcohol consumption per continent for every column
5. Print the mean, min and max values for spirit consumption


In [ ]:
url = "https://raw.githubusercontent.com/justmarkham/DAT8/master/data/drinks.csv"
df = pd.read_csv(url)

In [ ]:
df

country  beer_servings  spirit_servings  wine_servings  \
0    Afghanistan              0                0              0   
1        Albania             89              132             54   
2        Algeria             25                0             14   
3        Andorra            245              138            312   
4         Angola            217               57             45   
..           ...            ...              ...            ...   
188    Venezuela            333              100              3   
189      Vietnam            111                2              1   
190        Yemen              6                0              0   
191       Zambia             32               19              4   
192     Zimbabwe             64               18              4   

     total_litres_of_pure_alcohol continent  
0                             0.0        AS  
1                             4.9        EU  
2                             0.7        AF  
3                            12.4        EU  
4                             5.9        AF  
..                            ...       ...  
188                           7.7        SA  
189                           2.0        AS  
190                           0.1        AS  
191                           2.5        AF  
192                           4.7        AF  

[193 rows x 6 columns]

In [ ]:
count_continent = df.groupby('continent')['beer_servings'].mean(numeric_only = True)
count_continent

continent
AF     61.471698
AS     37.045455
EU    193.777778
OC     89.687500
SA    175.083333
Name: beer_servings, dtype: float64

# Task
Summarize the findings from the analysis of wine and spirit consumption across different continents. Calculate the median alcohol consumption per continent for all numeric alcohol consumption columns. Also, calculate the mean, minimum, and maximum values for spirit consumption by continent, and descriptive statistics (count, mean, std, min, 25%, 50%, 75%, max) for wine consumption by continent, using the dataframe `df`.

## Calculate Wine Consumption Statistics by Continent

### Subtask:
Group the DataFrame by 'continent' and calculate descriptive statistics (count, mean, std, min, 25%, 50%, 75%, max) for 'wine_servings'.


**Reasoning**:
To calculate the descriptive statistics for 'wine_servings' grouped by 'continent', I will use the `groupby()` method on the 'continent' column, select 'wine_servings', and then apply the `describe()` method.



In [ ]:
wine_stats = df.groupby('continent')['wine_servings'].describe()
wine_stats

count        mean        std  min   25%    50%     75%    max
continent                                                               
AF          53.0   16.264151  38.846419  0.0   1.0    2.0   13.00  233.0
AS          44.0    9.068182  21.667034  0.0   0.0    1.0    8.00  123.0
EU          45.0  142.222222  97.421738  0.0  59.0  128.0  195.00  370.0
OC          16.0   35.625000  64.555790  0.0   1.0    8.5   23.25  212.0
SA          12.0   62.416667  88.620189  1.0   3.0   12.0   98.50  221.0

In [ ]:
df.groupby('continent').mean(numeric_only = True)

beer_servings  spirit_servings  wine_servings  \
continent                                                  
AF             61.471698        16.339623      16.264151   
AS             37.045455        60.840909       9.068182   
EU            193.777778       132.555556     142.222222   
OC             89.687500        58.437500      35.625000   
SA            175.083333       114.750000      62.416667   

           total_litres_of_pure_alcohol  
continent                                
AF                             3.007547  
AS                             2.170455  
EU                             8.617778  
OC                             3.381250  
SA                             6.308333

In [ ]:
df.groupby('continent').median.
3(numeric_only = True)

beer_servings  spirit_servings  wine_servings  \
continent                                                  
AF                  32.0              3.0            2.0   
AS                  17.5             16.0            1.0   
EU                 219.0            122.0          128.0   
OC                  52.5             37.0            8.5   
SA                 162.5            108.5           12.0   

           total_litres_of_pure_alcohol  
continent                                
AF                                 2.30  
AS                                 1.20  
EU                                10.00  
OC                                 1.75  
SA                                 6.85

**Reasoning**:
The previous subtask of calculating descriptive statistics for 'wine_servings' by continent has been completed. Now, I will calculate the median alcohol consumption for each continent for all relevant alcohol consumption columns (beer_servings, spirit_servings, wine_servings, total_litres_of_pure_alcohol).



In [ ]:
median_alcohol_consumption = df.groupby('continent')[['beer_servings', 'spirit_servings', 'wine_servings', 'total_litres_of_pure_alcohol']].median()
median_alcohol_consumption

beer_servings  spirit_servings  wine_servings  \
continent                                                  
AF                  32.0              3.0            2.0   
AS                  17.5             16.0            1.0   
EU                 219.0            122.0          128.0   
OC                  52.5             37.0            8.5   
SA                 162.5            108.5           12.0   

           total_litres_of_pure_alcohol  
continent                                
AF                                 2.30  
AS                                 1.20  
EU                                10.00  
OC                                 1.75  
SA                                 6.85

**Reasoning**:
I need to calculate the mean, minimum, and maximum values for spirit consumption by continent. I will achieve this by grouping the DataFrame by 'continent', selecting the 'spirit_servings' column, and then applying the `agg` function with 'mean', 'min', and 'max'.



In [ ]:
spirit_stats_agg = df.groupby('continent')['spirit_servings'].agg(['mean', 'min', 'max'])
spirit_stats_agg

mean  min  max
continent                      
AF          16.339623    0  152
AS          60.840909    0  326
EU         132.555556    0  373
OC          58.437500    0  254
SA         114.750000   25  302

## Final Task

### Subtask:
Summarize the findings from the analysis of wine and spirit consumption across different continents.


## Summary:

### Data Analysis Key Findings

*   **Wine Consumption:** Europe (EU) exhibits the highest mean wine consumption at 142.22 servings, significantly higher than African and Asian continents, indicating a strong wine culture.
*   **Overall Median Consumption:** Europe consistently leads in median alcohol consumption across all categories: 219.0 servings for beer, 122.0 servings for spirits, 128.0 servings for wine, and 10.0 total litres of pure alcohol, highlighting its generally higher consumption levels.
*   **Spirit Consumption:** Europe also records the highest mean spirit consumption at 132.56 servings, with a maximum of 373 servings. South America follows with a notable mean of 114.75 servings.

### Insights or Next Steps

*   The findings suggest a clear geographical disparity in alcohol consumption patterns, with European countries generally having higher consumption across wine, spirits, and beer.
*   Further analysis could explore the cultural, economic, and historical factors contributing to these distinct consumption patterns across continents, potentially delving into specific country-level data within the higher-consuming continents like Europe and South America.
